File: apps/backend-python/src/services/auth_service.py
from __future__ import annotations
import secrets
from datetime import datetime, timedelta
from nanoid import generate
from sqlalchemy.orm import Session
from src.db.models import User, Auth, ApiKey, UserSession
from src.utils.hash import hash_password, verify_password
from src.utils.jwt import generate_refresh_token, sign_access_token
from src.utils.errors import BadRequestError, UnauthorizedError, InternalServerError
from sqlalchemy.exc import SQLAlchemyError

AUTH_GRACE_PERIOD_SECONDS = 5

class AuthService:
    def _create_tokens_and_session(self, db: Session, user: User, existing_session: UserSession = None) -> dict:
        """Internal helper function to create tokens and store sessions"""
        access_token = sign_access_token({
            "id": user.id, "email": user.email,
            "name": user.name, "role": user.role,
        })
        
        refresh_token = generate_refresh_token()
        hashed_rt = hash_password(refresh_token) 
        
        expires_at = datetime.utcnow() + timedelta(days=7) 
        now = datetime.utcnow()
        
        if existing_session:
            session_id = existing_session.id
            existing_session.hashed_refresh_token = hashed_rt
            existing_session.expires_at = expires_at
            existing_session.updated_at = now
        else:
            session_id = generate()
            db_session = UserSession(
                id=session_id,
                user_id=user.id,
                hashed_refresh_token=hashed_rt,
                expires_at=expires_at,
                is_valid=True,
                created_at=now,
                updated_at=now
            )
            db.add(db_session)
            
        return {
            "access_token": access_token, 
            "refresh_token": f"{session_id}::{refresh_token}"
        }

    def signup(self, db: Session, email: str, password: str, name: str) -> dict:
        existing = db.query(Auth).filter(Auth.email == email).first()
        if existing:
            raise BadRequestError("Email already registered")

        hashed = hash_password(password)
        user_id = generate()
        now = datetime.utcnow()

        user_count = db.query(User).count()
        role = "admin" if user_count == 0 else "user"

        user = User(
            id=user_id, email=email, name=name, role=role,
            profile_image_url="/user.png",
            last_active_at=now, created_at=now, updated_at=now,
        )
        auth = Auth(id=user_id, email=email, password=hashed, active=True)

        try:
            db.add(user)
            db.add(auth)
            
            tokens = self._create_tokens_and_session(db, user) # There is a db.add() call here
            
            db.commit()
            db.refresh(user)
            
            return {"user": user, **tokens}
        except SQLAlchemyError:
            db.rollback() 
            raise InternalServerError("Failed to register user")

    def signin(self, db: Session, email: str, password: str) -> dict:
        auth_record = db.query(Auth).filter(Auth.email == email).first()
        if not auth_record or not auth_record.active:
            raise UnauthorizedError("Invalid credentials")

        if not verify_password(password, auth_record.password):
            raise UnauthorizedError("Invalid credentials")

        user = db.query(User).filter(User.id == auth_record.id).first()
        if not user:
            raise UnauthorizedError("User not found")

        try:
            # Soft delete all expired user session
            db.query(UserSession).filter(
                UserSession.user_id == user.id,
                UserSession.expires_at < datetime.utcnow()
            ).update({"is_valid": False, "updated_at": datetime.utcnow()})

            tokens = self._create_tokens_and_session(db, user)
            db.commit() 
            return {"user": user, **tokens}
        except SQLAlchemyError:
            db.rollback()
            raise InternalServerError("Failed to create login session")

    def refresh_session(self, db: Session, raw_refresh_token: str) -> dict:
        try:
            session_id, plain_token = raw_refresh_token.split("::")
        except ValueError:
            raise UnauthorizedError("Invalid token format")

        db_session = (
                db.query(UserSession)
                .with_for_update() # Row-level locking to prevent race condition
                .filter(UserSession.id == session_id)
                .first()
            )
        
        if not db_session or not db_session.is_valid:
            raise UnauthorizedError("Session not found or revoked")
            
        now = datetime.utcnow()

        if db_session.expires_at < now:
            db_session.is_valid = False
            db_session.updated_at = now
            db.commit()
            raise UnauthorizedError("Session expired")

        if not verify_password(plain_token, db_session.hashed_refresh_token):
            raise UnauthorizedError("Invalid refresh token")
        
        # Grace period (5 seconds) to prevent token re-creation and DB updates
        if db_session.updated_at and (now - db_session.updated_at < timedelta(seconds=AUTH_GRACE_PERIOD_SECONDS)):
            return {"grace_period_active": True}

        user = db.query(User).filter(User.id == db_session.user_id).first()
        
        try:
            # Create or update refresh token in DB
            tokens = self._create_tokens_and_session(db, user, existing_session=db_session)
            db.commit()
            return {"user": user, **tokens}
        except SQLAlchemyError:
            db.rollback() 
            raise InternalServerError("Failed to refresh and rotate session")

    def signout(self, db: Session, raw_refresh_token: str | None) -> None:
        if not raw_refresh_token:
            return 
            
        try:
            session_id, _ = raw_refresh_token.split("::")
            # Soft delete current session
            db.query(UserSession).filter(UserSession.id == session_id).update({
                "is_valid": False,
                "updated_at": datetime.utcnow()
            })
            db.commit()
        except Exception:
            db.rollback()
            pass # 'Pass' because failure to delete shouldn't stop cookies deletion

    def generate_api_key(self, db: Session, user_id: str) -> str:
        key = f"sk-{secrets.token_urlsafe(36)}"
        now = datetime.utcnow()
        db.query(ApiKey).filter(ApiKey.user_id == user_id).delete()
        db.add(ApiKey(
            id=f"key_{user_id}", user_id=user_id, key=key,
            created_at=now, updated_at=now,
        ))
        db.commit()
        return key

    def delete_api_key(self, db: Session, user_id: str) -> None:
        db.query(ApiKey).filter(ApiKey.user_id == user_id).delete()
        db.commit()

    def get_user_by_api_key(self, db: Session, key: str) -> User | None:
        api_key = db.query(ApiKey).filter(ApiKey.key == key).first()
        if not api_key:
            return None
        return db.query(User).filter(User.id == api_key.user_id).first()


auth_service = AuthService()

File: apps/backend/src/services/auth.service.ts
import { nanoid } from 'nanoid';
import { eq } from 'drizzle-orm';
import { db } from '../config/database';
import { users, auths, apiKeys } from '../db/schema';
import { hashPassword, verifyPassword } from '../utils/hash';
import { signToken } from '../utils/jwt';
import { BadRequestError, UnauthorizedError } from '../utils/errors';
import { settingsService } from './settings.service';

export class AuthService {
  async signup(email: string, password: string, name: string) {
    
    const existing = await db.select().from(auths).where(eq(auths.email, email)).limit(1);
    if (existing.length > 0) {
      throw new BadRequestError('Email already registered');
    }

    
    const hashedPassword = await hashPassword(password);

    
    const userId = nanoid();
    const now = new Date();

    
    const userCount = await db.select().from(users);
    let role = 'user';

    if (userCount.length === 0) {
      role = 'admin';
    } else {
      
      const authSettings = await settingsService.getSettingsByCategory('auth');
      
      role = authSettings?.default_user_role || 'user';
    }

    
    await db.insert(users).values({
      id: userId,
      email,
      name,
      role,
      profileImageUrl: '/user.png',
      lastActiveAt: now,
      createdAt: now,
      updatedAt: now,
    });

    
    await db.insert(auths).values({
      id: userId,
      email,
      password: hashedPassword,
      active: true,
    });

    
    const token = signToken({ id: userId, email, name, role });

    const user = await db.select().from(users).where(eq(users.id, userId)).limit(1);

    return { user: user[0], token };
  }

  async signin(email: string, password: string) {
    
    const authRecord = await db.select().from(auths)
      .where(eq(auths.email, email))
      .limit(1);

    if (authRecord.length === 0 || !authRecord[0].active) {
      throw new UnauthorizedError('Invalid credentials');
    }

    
    const valid = await verifyPassword(password, authRecord[0].password);
    if (!valid) {
      throw new UnauthorizedError('Invalid credentials');
    }

    
    const user = await db.select().from(users)
      .where(eq(users.id, authRecord[0].id))
      .limit(1);

    if (user.length === 0) {
      throw new UnauthorizedError('User not found');
    }

    
    const token = signToken({
      id: user[0].id,
      email: user[0].email,
      name: user[0].name,
      role: user[0].role
    });

    return { user: user[0], token };
  }

  async generateApiKey(userId: string) {
    const key = `sk-${nanoid(48)}`;
    const now = new Date();

    
    await db.delete(apiKeys).where(eq(apiKeys.userId, userId));

    
    await db.insert(apiKeys).values({
      id: `key_${userId}`,
      userId,
      key,
      createdAt: now,
      updatedAt: now,
    });

    return key;
  }

  async deleteApiKey(userId: string) {
    await db.delete(apiKeys).where(eq(apiKeys.userId, userId));
  }

  async getUserByApiKey(key: string) {
    const apiKey = await db.select().from(apiKeys)
      .where(eq(apiKeys.key, key))
      .limit(1);

    if (apiKey.length === 0) {
      return null;
    }

    const user = await db.select().from(users)
      .where(eq(users.id, apiKey[0].userId))
      .limit(1);

    return user.length > 0 ? user[0] : null;
  }
}

export const authService = new AuthService();
